src/main/scala/loamstream/model/execute/ExecutionState.scala

```diff
@@ -0,0 +1,302 @@
+   */
+  def updateJobs(): ExecutionState.JobStatuses = jobStateListBox.get { _ =>
+    TimeUtils.time(s"updateJobs()", debug(_)) {
+      val currentJobStatuses = jobStatuses
+      
+      val eligible = currentJobStatuses.readyToRun
+      
+      TimeUtils.time(s"startRunning()", debug(_)) {
+        startRunning(eligible.keys)
+      }
+      
+      val toCancel = currentJobStatuses.cannotRun.keys
+      
+      TimeUtils.time(s"markAs(CouldNotStart)", debug(_)) { 
+        markAs(toCancel, JobStatus.CouldNotStart)
+      }
+      
+      currentJobStatuses
+    }
+  }
+
+  /**
+   * Returns a view of the current state of all jobs, useful to Executers.  Notably, jobs that are now
+   * runnable, and those that are now known to NEVER be runnable (due to a failure of one of their dependencies,
+   * say) are returned.
+   */
+  def jobStatuses: ExecutionState.JobStatuses = { 
+    val jobStates = jobStateListBox.get(_.snapshot)      
+  
+    TimeUtils.time("Computing JobStatuses", debug(_)) {
+      val numRunning = jobStates.count(_.cell.isRunning)
+      val numFinished = jobStates.count(_.cell.isFinished)
+      
+      val z = ExecutionState.JobStatuses.empty.copy(numRunning = numRunning, numFinished = numFinished)
+      
+      jobStates.foldLeft(z) { (acc, jobState) =>
+        import jobState.{job, cell}
+        
+        def anyDepsStopExecution: Boolean = {
+          def depCells = cellsFor(jobStates)(job.dependencies.map(_.job)) 
+          
+          cell.notStarted && depCells.exists(_.canStopExecution)
+        }
+        
+        def canRun: Boolean = this.canRun(jobStates)(jobState)
+        
+        if(canRun) { acc.withRunnable(jobState) }
+        else if(anyDepsStopExecution) { acc.withCannotRun(jobState) }
+        else { acc }
+      }
+    }
+  }
+  
+  /**
+   * Obtains the ExecutionCells for a given bunch of jobs, given the states of all jobs.
+   * (Returns an array for fast iteration (profiling turned this up). 
+   */
+  private def cellsFor(jobStates: JobStateList)(jobs: Set[JobNode]): Array[ExecutionCell] = {
+    val indexes = jobs.iterator.map(_.job).map(index(_))
+    
+    val cells: Array[ExecutionCell] = Array.ofDim[ExecutionCell](jobs.size)
+    
+    indexes.map(i => jobStates(i).cell).copyToArray(cells)
+    
+    cells
+  }
+  
+  /**
+   * Can a job run, given its current state, and the states of all other jobs.
+   */
+  private[execute] def canRun(jobStates: JobStateList)(jobState: JobState): Boolean = {
+    import jobState.{job, cell}
+    
+    cell.notStarted && {
+      val deps = job.dependencies
+      
+      deps.isEmpty || {
+        val depCells = cellsFor(jobStates)(deps)
+        
+        depCells.forall(_.isTerminal) && !depCells.exists(_.canStopExecution)
+      }
+    }
+  }
+  
+  /**
+   * Mark a job as running - change its status to Running and increment its run count.
+   */
+  def startRunning(jobs: TraversableOnce[LJob]): Unit = transition(jobs, _.startRunning)
```

Comment:
Should this be private? It called from `updateJobs`, but takes an "arbitrary" list of jobs. I'd hate to think external code could just pass in a list of random jobs to a random `ExecutionState` and start borking things by accident?

Alternatively, if the jobs started should always be local to this instance, instead of passing them in, maybe it just gets the list itself from it's own job list? I can see why it isn't done that way now; presumably it takes time to get that list and that's why there's `updateJobs` doing the multi-step process of finding that list of jobs.

Reply:
> Should this be private?

This question warms my cold New England heart.  And yes, it should, or at least package-private.

> Alternatively, if the jobs started should always be local to this instance, instead of passing them in, maybe it just gets the list itself from it's own job list?

This is basically the case; see `jobStatuses` and `updateJobs()`.